### PR TITLE
minor inconsistency in Categorical.remove_categories error message

### LINKED ISSUE #28669: minor inconsistency in Categorical.remove_categories error message
mypy error: `pandas\core\arrays\categorical.py:1125:13: error: Incompatible types in assignment (expression has type "List[Any]", variable has type "Set[Any]")  [assignment]`

#### Code Sample, a copy-pastable example if possible

```python
Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> import pandas as pd
>>> pd.__version__
'0.26.0.dev0+537.gd2bc6a497.dirty'
>>>
>>> cat = pd.Categorical(["a", "b", "c", "a"], ordered=True)
>>>
>>> cat
[a, b, c, a]
Categories (3, object): [a < b < c]
>>>
>>> cat = cat.remove_categories(["c"])
>>> cat
[a, b, NaN, a]
Categories (2, object): [a < b]
>>>
>>> cat = cat.remove_categories(["c"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\pandas-simonjayhawkins\pandas\core\arrays\categorical.py", line 1130, in remove_categories
    raise ValueError(msg.format(not_included=not_included))
ValueError: removals must all be in old categories: {'c'}
>>>
>>> cat = cat.remove_categories(["c", np.nan])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\pandas-simonjayhawkins\pandas\core\arrays\categorical.py", line 1130, in remove_categories
    raise ValueError(msg.format(not_included=not_included))
ValueError: removals must all be in old categories: ['c']
>>> cat
[a, b, NaN, a]
Categories (2, object): [a < b]
>>>

```
#### Problem description

The error message shows invalid removals as a set or a list depending on whether the removal list contains a null value.

#### Expected Output

consistent message. preferably as a list.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : d2bc6a49778407a02f31c40408e8d62cb24e6b64
python           : 3.7.4.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 58 Stepping 9, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : None.None

pandas           : 0.26.0.dev0+537.gd2bc6a497.dirty
numpy            : 1.16.4
pytz             : 2019.2
dateutil         : 2.8.0
pip              : 19.2.2
setuptools       : 41.0.1
Cython           : 0.29.13
pytest           : 5.0.1
hypothesis       : 4.32.2
sphinx           : 1.8.5
blosc            : None
feather          : None
xlsxwriter       : 1.1.8
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.8.0
pandas_datareader: None
bs4              : 4.7.1
bottleneck       : 1.2.1
fastparquet      : 0.3.0
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.1.1
numexpr          : 2.7.0
odfpy            : None
openpyxl         : 2.6.2
pandas_gbq       : None
pyarrow          : 0.13.0
pytables         : None
s3fs             : 0.2.1
scipy            : 1.3.1
sqlalchemy       : 1.3.7
tables           : 3.5.2
xarray           : 0.12.1
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.1.8
>>>

</details>

### PR DESCRIPTION
**Pull request by punndcoder28**
This pull request fixes minor inconsistency in Categorical.remove_categories error message

- [x] closes #28669 
- [ ] tests added / passed
-  passes `black pandas`
-  passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
-  whatsnew entry

Changed the error message to show invalid removals as a set. Added tests for removal of null from the categories. Parameterized pytest.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Nitpick: in the future please give the PR an informative title.  Most reviewers won't know off the top of their heads what 28677 is about.

**Comment by punndcoder28** [comment]
Yes my bad @jbrockmendel. I realized it after open the PR sorry.

**Review by jreback — CHANGES_REQUESTED** [review]
can you add or update the tests on remove_categories to account for this change (since nothing broke we likely are not testing it enough)

**Comment by punndcoder28** [comment]
I had tests as it were suggested by @simonjayhawkins. But there were errors and I didn't know how to handle that. If I could get some guidance I will surely add those and commit the changes.

**Comment by punndcoder28** [comment]
If you could tell me what tests I should create I'll create those and commit @jreback.

**Comment by simonjayhawkins** [comment]
@punndcoder28  suitable tests would be similar to those in pandas/tests/arrays/categorical/test_api.py in  https://github.com/pandas-dev/pandas/pull/28339/commits/8298620a3a2e2c88b922b5da5b76609e62c5d852

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
@punndcoder28 Thanks for the PR. can you add a whatsnew entry as well.

**Comment by punndcoder28** [comment]
`assert [NaN, b, c, NaN]\nCategories (2, object): [b < c] is None`. This shouldn't cause an error right. As  `[b < c] is None` states the statement `assert res is None` should evaluate to True and should not raise an `AssertionError`. Am I understanding this the wrong way?

File: pandas/tests/arrays/categorical/test_api.py [hunk 1]
```diff
@@ -337,12 +338,31 @@ def test_remove_categories(self):
...
+        new = Categorical([np.nan, "b", "c", np.nan], categories=["b", "c"], ordered=True)
+
+        res = cat.remove_categories("a")
+        tm.assert_categorical_equal(cat, old)
+        tm.assert_categorical_equal(res, new)
+
+        res = cat.remove_categories(["a"])
+        tm.assert_categorical_equal(cat, old)
+        tm.assert_categorical_equal(res, new)
+
```
Comment (simonjayhawkins) [inline pandas/tests/arrays/categorical/test_api.py]: we're only dealing with the message in the Exception in this PR so remove these tests.

File: pandas/tests/arrays/categorical/test_api.py [hunk 2]
```diff
@@ -337,12 +338,31 @@ def test_remove_categories(self):
...
         assert res is None
 
         # removal is not in categories
-        with pytest.raises(ValueError):
+        msg = re.escape("removals must all be in old categories: ['c']")
+        with pytest.raises(ValueError, match=msg):
             cat.remove_categories(["c"])
 
+        msg = re.escape("removals must all be in old categories: [np.nan]")
```
Comment (simonjayhawkins) [inline pandas/tests/arrays/categorical/test_api.py]: i think we should have 4 tests. that all produce the same error message so only need to assign to msg once.

File: pandas/tests/arrays/categorical/test_api.py [hunk 3]
```diff
@@ -337,12 +338,31 @@ def test_remove_categories(self):
...
+        with pytest.raises(ValueError):
+            cat.remove_categories([np.nan])
```
Comment (simonjayhawkins) [inline pandas/tests/arrays/categorical/test_api.py]: I think the 4 test cases should be..

```
cat.remove_categories(["c"])
cat.remove_categories(["c", np.nan])
cat.remove_categories("c")
cat.remove_categories(["c","c" ])
```
 and all these cases should raise the same exception with the same message.

if you could split the pytest raises test from `test_remove_categories` into a new test `test_remove_categories_raises` and parameterize that would be great.

Comment (punndcoder28) [reply pandas/tests/arrays/categorical/test_api.py]: Shall I do it for all the above stated ones?

Comment (simonjayhawkins) [reply pandas/tests/arrays/categorical/test_api.py]: The first two are the ones in the OP #28669, so necessary to close issue.

The second two as suggested in https://github.com/pandas-dev/pandas/issues/28669#issuecomment-536309377

**Comment by punndcoder28** [comment]
Hey @simonjayhawkins. I've made the changes as requested. Could you check and let me know if any further changes are needed?

File: pandas/tests/arrays/categorical/test_api.py [hunk 4]
```diff
@@ -337,11 +337,28 @@ def test_remove_categories(self):
...
         res = cat.remove_categories("c", inplace=True)
         tm.assert_categorical_equal(cat, new)
+
         assert res is None
 
-        # removal is not in categories
-        with pytest.raises(ValueError):
-            cat.remove_categories(["c"])
+    @pytest.mark.parametrize(
+        "message, old, category", [
```
Comment (simonjayhawkins) [inline pandas/tests/arrays/categorical/test_api.py]: no need to parametrize `message` and `old`. and I would rename `category` to `removals` to be consistent with the argument name, see https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.CategoricalIndex.remove_categories.html

**Comment by simonjayhawkins** [comment]
@punndcoder28 Thanks for the updates. can you also add a what's new.

File: pandas/tests/arrays/categorical/test_api.py (hunk 4 above)
Comment (punndcoder28) [reply pandas/tests/arrays/categorical/test_api.py]: `old` is needed to create the categorical list `cat` and `msg` to check the error message.

Comment (simonjayhawkins) [reply pandas/tests/arrays/categorical/test_api.py]: the values for `old` and `msg` are the same for all the tests. so no need to parametrize.

**Comment by punndcoder28** [comment]
Hey @simonjayhawkins. I've made the requested changes can you check it out and let me know if you need any more changes.

File: pandas/tests/arrays/categorical/test_api.py [hunk 5]
```diff
@@ -337,11 +337,23 @@ def test_remove_categories(self):
         # inplace == True
         res = cat.remove_categories("c", inplace=True)
         tm.assert_categorical_equal(cat, new)
+
```
Comment (simonjayhawkins) [inline pandas/tests/arrays/categorical/test_api.py]: can you remove this added blank line

File: pandas/tests/arrays/categorical/test_api.py [hunk 6]
```diff
@@ -337,11 +337,23 @@ def test_remove_categories(self):
...
         assert res is None
 
-        # removal is not in categories
-        with pytest.raises(ValueError):
-            cat.remove_categories(["c"])
+    @pytest.mark.parametrize(
+        "removals", [
+            (["c"]),
```
Comment (simonjayhawkins) [inline pandas/tests/arrays/categorical/test_api.py]: for a single parameter, boxing the parameters in a tuple is not required.

File: pandas/tests/arrays/categorical/test_api.py [hunk 7]
```diff
@@ -337,11 +337,23 @@ def test_remove_categories(self):
...
+            (["c", np.nan]),
+            ("c"),
+            (["c", "c"])
+        ],
+    )
+    def test_remove_categories_raises(self, removals):
+        cat = Categorical(["a", "b", "a"])
+        message = "removals must all be in old categories: {'c'}"
```
Comment (simonjayhawkins) [inline pandas/tests/arrays/categorical/test_api.py]: for safety can you add the `re.escape` back here.

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
ci is failing, see https://dev.pandas.io/docs/development/contributing.html#python-pep8-black

and add a whatsnew, doc/source/whatsnew/v1.0.0.rst

File: pandas/tests/arrays/categorical/test_api.py (hunk 7 above)
Comment (punndcoder28) [reply pandas/tests/arrays/categorical/test_api.py]: I've added that. Thanks.

File: pandas/tests/arrays/categorical/test_api.py (hunk 3 above)
Comment (punndcoder28) [reply pandas/tests/arrays/categorical/test_api.py]: I've the all the cases.

File: doc/source/whatsnew/v1.0.0.rst [hunk 8]
```diff
@@ -166,7 +166,7 @@ Categorical
 
 - Added test to assert the :func:`fillna` raises the correct ValueError message when the value isn't a value from categories (:issue:`13628`)
 - Bug in :meth:`Categorical.astype` where ``NaN`` values were handled incorrectly when casting to int (:issue:`28406`)
--
+- Changed the error message in `remove_categories` to show the invalid removals as a set.
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.0.0.rst]:
```suggestion
- Changed the error message in :meth:`Categorical.remove_categories` to show the invalid removals as a set.
```

Comment (punndcoder28) [reply doc/source/whatsnew/v1.0.0.rst]: Thanks.

File: doc/source/whatsnew/v1.0.0.rst [hunk 9]
```diff
@@ -166,7 +166,7 @@ Categorical
 
 - Added test to assert the :func:`fillna` raises the correct ValueError message when the value isn't a value from categories (:issue:`13628`)
 - Bug in :meth:`Categorical.astype` where ``NaN`` values were handled incorrectly when casting to int (:issue:`28406`)
--
+- Changed the error message in :meth:`Categorical.remove_categories` to show the invalid removals as a set.
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.0.0.rst]:
```suggestion
- Changed the error message in :meth:`Categorical.remove_categories` to always show the invalid removals as a set (:issue:`28669`)
```

File: pandas/tests/arrays/categorical/test_api.py [hunk 10]
```diff
@@ -339,9 +340,13 @@ def test_remove_categories(self):
         tm.assert_categorical_equal(cat, new)
         assert res is None
 
-        # removal is not in categories
-        with pytest.raises(ValueError):
-            cat.remove_categories(["c"])
+    @pytest.mark.parametrize("removals", [["c"], (["c", np.nan]), ("c"), (["c", "c"])])
```
Comment (simonjayhawkins) [inline pandas/tests/arrays/categorical/test_api.py]:
```suggestion
    @pytest.mark.parametrize("removals", [["c"], ["c", np.nan], "c", ["c", "c"]])
```

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
@punndcoder28 generally looks good. a couple of small changes. also need to merge master to resolve merge conflict. ping on green.

**Comment by punndcoder28** [comment]
> @punndcoder28 generally looks good. a couple of small changes. also need to merge master to resolve merge conflict. ping on green.

How do I resolve a conflict? Or will it be done by the panda guys?

**Comment by simonjayhawkins** [comment]
> How do I resolve a conflict? I tried doing it manually but didn't work.

see https://dev.pandas.io/docs/development/contributing.html#updating-your-pull-request
```
git fetch upstream
git merge upstream/master
```

File: doc/source/whatsnew/v1.0.0.rst [hunk 11]
```diff
@@ -166,8 +166,10 @@ Categorical
 
 - Added test to assert the :func:`fillna` raises the correct ValueError message when the value isn't a value from categories (:issue:`13628`)
 - Bug in :meth:`Categorical.astype` where ``NaN`` values were handled incorrectly when casting to int (:issue:`28406`)
--
--
+- :meth:`Categorical.searchsorted` and :meth:`CategoricalIndex.searchsorted` now work on unordered categoricals also (:issue:`21667`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: looks like you picked up some other issue what's news here, can you edit to only add yours

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by punndcoder28** [comment]
I did what was mentioned here 

> ci is failing, see https://dev.pandas.io/docs/development/contributing.html#python-pep8-black

But still the tests are failing.

**Comment by jbrockmendel** [comment]
You need to run `isort pandas/tests/arrays/categorical/test_api.py`

**Review by simonjayhawkins — APPROVED** [review]
@punndcoder28 lgtm. @jreback

**Comment by WillAyd** [comment]
Hmm is this actually changing anything on master? I ran one of the test cases but already see the intended result?

```python
>>> pd.Categorical(["a", "b", "a"]).remove_categories(["c", "c"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/williamayd/clones/pandas/pandas/core/arrays/categorical.py", line 1132, in remove_categories
    raise ValueError(msg.format(not_included=not_included))
ValueError: removals must all be in old categories: {'c'}
```

**Comment by simonjayhawkins** [comment]
it's changeing the output of `cat = cat.remove_categories(["c", np.nan])`

from `ValueError: removals must all be in old categories: ['c']`

to `ValueError: removals must all be in old categories: {'c'}`

for all other cases the message remain unchanged and tests are added for these. i.e.

`cat = cat.remove_categories(["c"])`
`cat = cat.remove_categories("c")`
`cat = cat.remove_categories(["c", "c"])`

**Review by WillAyd — APPROVED** [review]
Ah OK thanks for clarifying @simonjayhawkins

**Comment by WillAyd** [comment]
Thanks @punndcoder28

**Pull request merged**